File: tempestweb/native/gamepad.py
"""Native gamepad capability over the Gamepad Web API.

:func:`state` sends a ``gamepad.state`` ``native_call``; ``client/native/
gamepad.js`` reads ``navigator.getGamepads()`` and returns a snapshot of each
connected gamepad's axes and buttons. The gamepad shape is browser-defined, so
snapshots pass through as ``dict[str, Any]`` rather than being modeled as
dataclasses.
"""

from __future__ import annotations

from collections.abc import AsyncIterator
from typing import Any, cast

from tempestweb.native.dispatch import native_events, send_native_call

__all__ = ["state", "watch"]


async def state() -> list[dict[str, Any]]:
    """Snapshot the currently connected gamepads.

    Returns:
        A list of gamepad snapshots (axes, buttons, id, …) as JSON-able dicts; an
        empty list when no gamepad is connected.

    Raises:
        NativeError: If the API is unavailable (``unavailable``).
        BrowserUnavailableError: If called with no native bridge installed.
    """
    value = await send_native_call("gamepad.state", {})
    return cast("list[dict[str, Any]]", value.get("gamepads", []))


async def watch() -> AsyncIterator[list[dict[str, Any]]]:
    """Stream gamepad snapshots from the browser (event channel / T-EV).

    Yields a fresh list of gamepad snapshots on every animation-frame poll (or on
    connect/disconnect) until the ``async for`` loop is exited (which closes the
    subscription). Consume it with::

        async for gamepads in native.gamepad.watch():
            app.set_state(lambda s: setattr(s, "gamepads", gamepads))

    Yields:
        Each snapshot: a list of connected gamepads as JSON-able dicts (empty when
        none are connected).

    Raises:
        NativeError: If the browser reports the subscription failed.
        BrowserUnavailableError: If no bridge is installed, or the installed bridge
            does not support the event channel.
    """
    async for value in native_events("gamepad.watch", {}):
        yield cast("list[dict[str, Any]]", value["gamepads"])

File: client/native/gamepad.js
// native/gamepad.js — Gamepad API glue for the Tier-3 seam.
//
// `navigator.getGamepads()` returns a snapshot (with null holes for empty slots)
// of the connected gamepads' button/axis state. We drop the holes and serialize
// each pad's buttons and axes into a JSON-able shape.

import { CapabilityError } from "./index.js";

/**
 * Snapshot the connected gamepads' button and axis state.
 * @param {Object} _args
 * @param {import("./index.js").NativeDeps} deps
 * @returns {Promise<{gamepads:Array<Object>}>}
 * @throws {CapabilityError} unavailable.
 */
export async function gamepadState(_args, deps) {
  const nav = deps.navigator || /** @type {any} */ (globalThis).navigator;
  if (!nav || typeof nav.getGamepads !== "function") {
    throw new CapabilityError("unavailable", "the Gamepad API is not available");
  }
  const pads = Array.from(nav.getGamepads()).filter((g) => g);
  const gamepads = pads.map((g) => ({
    index: g.index,
    id: g.id,
    buttons: Array.from(g.buttons).map((b) => ({ pressed: b.pressed, value: b.value })),
    axes: Array.from(g.axes),
  }));
  return { gamepads };
}
